### PR TITLE
Ensure failoverCandidates have no servers not in Plan.

### PR DESCRIPTION
**Pull request by neunhoef**
Fix failoverCandidates for servers not in Plan.

Problem: we must never fail over to a server which is not in Plan
because it might not have the data.

ToDo:
  - [X] AddFollower, FailedFollower, CleanOutServer and MoveShard will not
        add a server to the Plan which is in failoverCandidates for this
        shard or any of the ones with distributeShardsLike.
  - [x] add preconditions to ensure that it is snapshot-safe
  - [x] make dbserver Maintenance remove servers that are not in Plan from
        failoverCandidates.

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12166/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12167/

File: arangod/Agency/MoveShard.cpp [hunk 1]
```diff
@@ -298,6 +299,19 @@ bool MoveShard::start(bool&) {
...
 
+  // Compute group to move shards together:
+  std::vector<Job::shard_t> shardsLikeMe =
+      clones(_snapshot, _database, _collection, _shard);
+
+  if (found < 0) {   // not in Plan, then it must not be a failoverCandidate:
+    auto failoverCands = Job::findAllFailoverCandidates(
+        _snapshot, _database, shardsLikeMe);
+    if (failoverCands.find(_to) != failoverCands.end()) {
+      finish("", "", false, "toServer must not be in failoverCandidates for shard or any of its distributeShardsLike colleagues");
```
Comment (maierlars) [inline arangod/Agency/MoveShard.cpp:310]: Should this finish the job or just wait and do nothing?

**Review by kvahed — APPROVED** [review]
LGTM

**Review by maierlars — APPROVED** [review]

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12169/

File: arangod/Agency/MoveShard.cpp (hunk 1 above)
Comment (neunhoef) [reply arangod/Agency/MoveShard.cpp:310]: That is a good question. I do not like stuck jobs. Therefore I opted for abortion. But I can understand the argument that the client might be happier if we wait until the server is out of `failoverCandidates` and then do his bidding.

Comment (neunhoef) [reply arangod/Agency/MoveShard.cpp:310]: In all likelyhood, this will not make a difference in real life...

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/13787/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12172/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12174/

**Comment by KVS85** [comment]
```
[ 80%] Building CXX object arangod/CMakeFiles/arango_aql.dir/Aql/tokens.cpp.o
Aql/tokens.cpp:997:2: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
        register yy_state_type yy_current_state;
```

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-mac-onlycatch/9622/

**Comment by KVS85** [comment]
Tests blue.

**Review by mchacki — APPROVED** [review]
LGTM

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12182/

**Comment by KVS85** [comment]
Tests blue except known failure BTS-182 on Windows.

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12197/

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12202/

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12206/

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12211/

File: arangod/Agency/FailedFollower.cpp [hunk 2]
```diff
@@ -259,7 +276,22 @@ bool FailedFollower::start(bool& aborts) {
...
+            [this, &job](Slice plan, Slice current,
+                             std::string& planPath,
+                             std::string& curPath) {
+              // take off "servers" from curPath and add
+              // "failoverCandidates":
+              std::string foCandsPath = curPath.substr(0, curPath.size() - 7);
+              foCandsPath += StaticStrings::FailoverCandidates;
+              auto foCands = this->_snapshot.hasAsSlice(foCandsPath);
+              if (foCands.second) {
+                addPreconditionUnchanged(job, foCandsPath, foCands.first);
```
Comment (jsteemann) [inline arangod/Agency/FailedFollower.cpp:292]: Is "failoverCandidates" a mandatory attribute?
If not, and if it doesn't yet exist, we may end up with no precondition here. In which case we may overwrite changes to the same key by some other concurrent operation?
Maybe this is not possible in reality, but it should be checked.

File: arangod/Agency/FailedFollower.cpp [hunk 3]
```diff
@@ -166,8 +166,25 @@ bool FailedFollower::start(bool& aborts) {
     return false;
   }
 
-  // Get proper replacement
-  _to = randomIdleAvailableServer(_snapshot, planned);
+  // Exclude servers in failoverCandidates for some clone and those in Plan:
```
Comment (jsteemann) [inline arangod/Agency/FailedFollower.cpp:169]: the comment here says we are excluding servers in failoverCandidates, but line 177 looks to me as we would be excluding servers that are _not_ in failoverCandidates. is this intentional?

File: arangod/Agency/FailedLeader.cpp [hunk 4]
```diff
@@ -238,8 +238,25 @@ bool FailedLeader::start(bool& aborts) {
...
 
+  // Exclude servers in failoverCandidates for some clone and those in Plan:
+  auto shardsLikeMe = clones(_snapshot, _database, _collection, _shard);
+  auto failoverCands = Job::findAllFailoverCandidates(
+      _snapshot, _database, shardsLikeMe);
+  std::vector<std::string> excludes;
+  for (const auto& s : VPackArrayIterator(planned)) {
+    if (s.isString()) {
+      std::string id = s.copyString();
+      if (failoverCands.find(id) == failoverCands.end()) {
```
Comment (jsteemann) [inline arangod/Agency/FailedLeader.cpp:249]: the comment in line 241 says we are excluding servers in failoverCandidates, but line 249 looks to me as if we would be excluding servers that are _not_ in failoverCandidates. is this intentional?

File: arangod/Agency/FailedLeader.cpp [hunk 5]
```diff
@@ -309,13 +326,22 @@ bool FailedLeader::start(bool& aborts) {
...
+                             std::string& planPath,
+                             std::string& curPath) {
+              addPreconditionUnchanged(pending, curPath, current);
+              // take off "servers" from curPath and add
+              // "failoverCandidates":
+              std::string foCandsPath = curPath.substr(0, curPath.size() - 7);
+              foCandsPath += StaticStrings::FailoverCandidates;
+              auto foCands = this->_snapshot.hasAsSlice(foCandsPath);
+              if (foCands.second) {
+                addPreconditionUnchanged(pending, foCandsPath, foCands.first);
```
Comment (jsteemann) [inline arangod/Agency/FailedLeader.cpp:342]: same here, don't we need to add a precondition in case there have been no failoverCandidates yet?

**Comment by KVS85** [comment]
Tests blue.

File: arangod/Agency/MoveShard.cpp [hunk 6]
```diff
@@ -392,6 +404,20 @@ bool MoveShard::start(bool&) {
...
+          [this, &pending](Slice plan, Slice current,
+                           std::string& planPath,
+                           std::string& curPath) {
+            // take off "servers" from curPath and add
+            // "failoverCandidates":
+            std::string foCandsPath = curPath.substr(0, curPath.size() - 7);
+            foCandsPath += StaticStrings::FailoverCandidates;
+            auto foCands = this->_snapshot.hasAsSlice(foCandsPath);
+            if (foCands.second) {
+              addPreconditionUnchanged(pending, foCandsPath, foCands.first);
```
Comment (jsteemann) [inline arangod/Agency/MoveShard.cpp:418]: same here... no precondition is set when there is no "failoverCandidates" attribute yet.

File: arangod/Cluster/FollowerInfo.cpp [hunk 7]
```diff
@@ -162,28 +162,39 @@ Result FollowerInfo::remove(ServerID const& sid) {
...
                        sid) != _failoverCandidates->end());
   auto oldFollowers = _followers;
   auto oldFailovers = _failoverCandidates;
   {
     auto v = std::make_shared<std::vector<ServerID>>();
-    TRI_ASSERT(!_followers->empty());  // well we found the element above \o/
-    v->reserve(_followers->size() - 1);
+    if (!_followers->empty()) {
+      v->reserve(_followers->size() - 1);
+    }
```
Comment (jsteemann) [inline arangod/Cluster/FollowerInfo.cpp:188]: I vote for removing the if ... reserve altogether. It is a micro optimization that adds complexity and already has led to a problem before. Not reserving space will not kill us performance-wise, as we are talking about 1 or 2 followers only.

File: arangod/Cluster/FollowerInfo.cpp [hunk 8]
```diff
@@ -162,28 +162,39 @@ Result FollowerInfo::remove(ServerID const& sid) {
...
                      std::back_inserter(*v.get()), sid);
     _followers = v;  // will cast to std::vector<ServerID> const
   }
   {
     auto v = std::make_shared<std::vector<ServerID>>();
-    TRI_ASSERT(!_failoverCandidates->empty());  // well we found the element above \o/
-    v->reserve(_failoverCandidates->size() - 1);
+    if (!_failoverCandidates->empty()) {
+      v->reserve(_failoverCandidates->size() - 1);
+    }
```
Comment (jsteemann) [inline arangod/Cluster/FollowerInfo.cpp:197]: Same here. Let's remove this.

File: arangod/Agency/FailedFollower.cpp (hunk 2 above)
Comment (maierlars) [reply arangod/Agency/FailedFollower.cpp:292]: It should be asserted, that `foCands.second` is true. This attribute is mandatory afaik.

File: arangod/Cluster/FollowerInfo.cpp (hunk 7 above)
Comment (mchacki) [reply arangod/Cluster/FollowerInfo.cpp:188]: I agree the resizing here does not really help, and can be removed.

**Review by jsteemann — APPROVED** [review]

**Pull request merged**